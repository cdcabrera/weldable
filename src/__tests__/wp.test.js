const path = require('path');
const crypto = require('crypto');
const wp = require('../wp');
const { OPTIONS } = require('../global');

describe('webpack', () => {
  const baseOptions = {
    dotenv: {
      NODE_ENV: 'development',
      _BUILD_DIST_DIR: path.join(fixturePath, 'dist'),
      _BUILD_HOST: 'localhost',
      _BUILD_PORT: 3000,
      _BUILD_RELATIVE_DIRNAME: fixturePath,
      _BUILD_SRC_DIR: path.join(fixturePath, 'src')
    }
  };

  const cleanConfig = (obj, isHash) => {
    const contents = JSON.stringify(obj, null, 2).replace(/"[a-z0-9/-_.,*()\s]*\/weldable\//gi, '"./');
    if (isHash) {
      return crypto.createHash('md5').update(contents).digest('hex');
    }

    return contents;
  };

  it('should return specific properties', () => {
    expect(wp).toMatchSnapshot('specific properties');
  });

  it('should create a basic webpack config', async () => {
    const { mockClear } = mockObjectProperty(OPTIONS, {
      dotenv: {
        ...baseOptions.dotenv
      }
    });

    const dev = await wp.createWpConfig({
      nodeEnv: 'development'
    });

    const prod = await wp.createWpConfig();

    expect({
      dev: cleanConfig(dev),
      prod: cleanConfig(prod)
    }).toMatchSnapshot('basic configurations');
    mockClear();
  });

  it('should extend a basic webpack config using commonJS resources', async () => {
    const { mockClear } = mockObjectProperty(OPTIONS, {
      dotenv: {
        ...baseOptions.dotenv
      }
    });

    const commonDev = await wp.createWpConfig({
      nodeEnv: 'development',
      extendedConfigs: [path.resolve(fixturePath, 'webpack.customConfigDev.js')]
    });

    expect({
      common: cleanConfig(commonDev)
    }).toMatchSnapshot('configurations');
    mockClear();
  });

  it('should attempt to start a development server', async () => {
    const { mockClear } = mockObjectProperty(OPTIONS, {
      dotenv: {
        ...baseOptions.dotenv
      }
    });

    const basic = await wp.createWpConfig({ nodeEnv: 'development' });
    const startDevServer = jest.fn();
    const webpackDevServer = jest.fn().mockImplementation(() => ({ start: startDevServer }));
    await wp.startWp(basic, { nodeEnv: 'development' }, { WebpackDevServer: webpackDevServer });

    expect(webpackDevServer).toHaveBeenCalledTimes(1);
    expect(startDevServer).toHaveBeenCalledTimes(1);

    mockClear();
  });

  it('should attempt to produce build output bundle', async () => {
    const { mockClear } = mockObjectProperty(OPTIONS, {
      dotenv: {
        ...baseOptions.dotenv,
        NODE_ENV: 'production'
      }
    });

    const basic = await wp.createWpConfig({ nodeEnv: 'production' });
    const webpack = jest.fn();
    await wp.startWp(basic, { nodeEnv: 'production' }, { webpack });

    expect(webpack).toHaveBeenCalledTimes(1);
    mockClear();
  });

  it('should handle build output bundle errors, stats errors, and stat output', () => {
    const errorSpy = jest.spyOn(console, 'error').mockImplementation((...args) => args);
    const logSpy = jest.spyOn(console, 'log').mockImplementation((...args) => args);

    wp.startWpErrorStatsHandler(new Error('basic error'));
    wp.startWpErrorStatsHandler(undefined, {
      hasErrors: () => true,
      compilation: { errors: [new Error('compile error')] }
    });
    wp.startWpErrorStatsHandler(
      undefined,
      {
        hasErrors: () => false,
        toJson: () => ({ lorem: 'ipsum' }),
        toString: () => 'lorem: ipsum'
      },
      {
        statsFile: 'testJson',
        statsPath: tempFixturePath
      }
    );

    expect({
      errors: errorSpy.mock.calls,
      logs: logSpy.mock.calls
    }).toMatchSnapshot('startWpErrorStatsHandler');

    errorSpy.mockClear();
    logSpy.mockClear();
  });
});
