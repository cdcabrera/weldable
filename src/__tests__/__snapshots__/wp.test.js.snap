// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`webpack should create a basic webpack config: basic configurations 1`] = `
{
  "dev": "[
  {
    "context": "./__fixtures__",
    "entry": {
      "app": [
        "./__fixtures__/src/index.js"
      ]
    },
    "output": {
      "filename": "[name].bundle.js",
      "path": "./__fixtures__/dist",
      "clean": true
    },
    "module": {
      "rules": [
        {
          "test": {},
          "type": "asset/resource",
          "generator": {
            "filename": "fonts/[hash][ext][query]"
          }
        },
        {
          "test": {},
          "type": "asset",
          "parser": {
            "dataUrlCondition": {
              "maxSize": 5000
            }
          },
          "generator": {
            "filename": "images/[hash][ext][query]"
          }
        },
        {
          "test": {},
          "type": "asset",
          "parser": {
            "dataUrlCondition": {
              "maxSize": 5000
            }
          },
          "generator": {
            "filename": "images/[hash][ext][query]"
          }
        },
        {
          "test": {},
          "use": [
            "./node_modules/mini-css-extract-plugin/dist/loader.js",
            "css-loader"
          ]
        }
      ]
    },
    "plugins": [
      {
        "config": {
          "path": "./__fixtures__/.env.local",
          "prefix": "process.env.",
          "systemvars": true,
          "silent": true
        }
      },
      {
        "config": {
          "path": "./__fixtures__/.env",
          "prefix": "process.env.",
          "systemvars": true,
          "silent": true
        }
      },
      {
        "userOptions": {},
        "version": 5,
        "options": {
          "template": "auto",
          "templateContent": false,
          "filename": "index.html",
          "publicPath": "auto",
          "hash": false,
          "inject": "head",
          "scriptLoading": "defer",
          "compile": true,
          "favicon": false,
          "minify": "auto",
          "cache": true,
          "showErrors": true,
          "chunks": "all",
          "excludeChunks": [],
          "chunksSortMode": "auto",
          "meta": {},
          "base": false,
          "title": "Webpack App",
          "xhtml": false
        }
      },
      {
        "config": {
          "path": "./__fixtures__/.env.development.local",
          "prefix": "process.env.",
          "systemvars": true,
          "silent": true
        }
      },
      {
        "config": {
          "path": "./__fixtures__/.env.development",
          "prefix": "process.env.",
          "systemvars": true,
          "silent": true
        }
      },
      {
        "config": {
          "path": "./__fixtures__/.env.local",
          "prefix": "process.env.",
          "systemvars": true,
          "silent": true
        }
      },
      {
        "config": {
          "path": "./__fixtures__/.env",
          "prefix": "process.env.",
          "systemvars": true,
          "silent": true
        }
      },
      {
        "_sortedModulesCache": {},
        "options": {
          "filename": "[name].bundle.css",
          "ignoreOrder": false,
          "runtime": true,
          "chunkFilename": "[name].bundle.css"
        },
        "runtimeOptions": {
          "linkType": "text/css"
        }
      }
    ],
    "resolve": {
      "symlinks": false,
      "cacheWithContext": false
    },
    "mode": "development",
    "devtool": "eval-source-map",
    "devServer": {
      "host": "localhost",
      "port": 3000,
      "compress": true,
      "historyApiFallback": true,
      "hot": true,
      "devMiddleware": {
        "stats": "errors-warnings",
        "writeToDisk": false
      },
      "client": {
        "overlay": false,
        "progress": true
      },
      "static": {
        "directory": "./__fixtures__/dist"
      },
      "watchFiles": {
        "paths": [
          "src/**/*"
        ]
      }
    }
  }
]",
  "prod": "[
  {
    "context": "./__fixtures__",
    "entry": {
      "app": [
        "./__fixtures__/src/index.js"
      ]
    },
    "output": {
      "filename": "[name].[contenthash:8].js",
      "path": "./__fixtures__/dist",
      "clean": true,
      "chunkFilename": "[name].[contenthash:8].chunk.js"
    },
    "module": {
      "rules": [
        {
          "test": {},
          "type": "asset/resource",
          "generator": {
            "filename": "fonts/[hash][ext][query]"
          }
        },
        {
          "test": {},
          "type": "asset",
          "parser": {
            "dataUrlCondition": {
              "maxSize": 5000
            }
          },
          "generator": {
            "filename": "images/[hash][ext][query]"
          }
        },
        {
          "test": {},
          "type": "asset",
          "parser": {
            "dataUrlCondition": {
              "maxSize": 5000
            }
          },
          "generator": {
            "filename": "images/[hash][ext][query]"
          }
        },
        {
          "test": {},
          "use": [
            "./node_modules/mini-css-extract-plugin/dist/loader.js",
            "css-loader"
          ]
        }
      ]
    },
    "plugins": [
      {
        "config": {
          "path": "./__fixtures__/.env.local",
          "prefix": "process.env.",
          "systemvars": true,
          "silent": true
        }
      },
      {
        "config": {
          "path": "./__fixtures__/.env",
          "prefix": "process.env.",
          "systemvars": true,
          "silent": true
        }
      },
      {
        "userOptions": {},
        "version": 5,
        "options": {
          "template": "auto",
          "templateContent": false,
          "filename": "index.html",
          "publicPath": "auto",
          "hash": false,
          "inject": "head",
          "scriptLoading": "defer",
          "compile": true,
          "favicon": false,
          "minify": "auto",
          "cache": true,
          "showErrors": true,
          "chunks": "all",
          "excludeChunks": [],
          "chunksSortMode": "auto",
          "meta": {},
          "base": false,
          "title": "Webpack App",
          "xhtml": false
        }
      },
      {
        "config": {
          "path": "./__fixtures__/.env.development.local",
          "prefix": "process.env.",
          "systemvars": true,
          "silent": true
        }
      },
      {
        "config": {
          "path": "./__fixtures__/.env.development",
          "prefix": "process.env.",
          "systemvars": true,
          "silent": true
        }
      },
      {
        "config": {
          "path": "./__fixtures__/.env.local",
          "prefix": "process.env.",
          "systemvars": true,
          "silent": true
        }
      },
      {
        "config": {
          "path": "./__fixtures__/.env",
          "prefix": "process.env.",
          "systemvars": true,
          "silent": true
        }
      },
      {
        "_sortedModulesCache": {},
        "options": {
          "filename": "[name].[contenthash:8].css",
          "ignoreOrder": false,
          "runtime": true,
          "chunkFilename": "[name].[contenthash:8].chunk.css"
        },
        "runtimeOptions": {
          "linkType": "text/css"
        }
      }
    ],
    "resolve": {
      "symlinks": false,
      "cacheWithContext": false
    },
    "mode": "development",
    "optimization": {
      "minimize": true,
      "minimizer": [
        {
          "options": {
            "test": {},
            "extractComments": true,
            "parallel": true,
            "minimizer": {
              "options": {}
            }
          }
        },
        {
          "options": {
            "test": {},
            "parallel": true,
            "minimizer": {
              "options": {
                "preset": [
                  "default",
                  {
                    "mergeLonghand": false
                  }
                ]
              }
            }
          }
        }
      ],
      "runtimeChunk": "single",
      "splitChunks": {
        "chunks": "all",
        "cacheGroups": {
          "vendor": {
            "chunks": "all",
            "name": "vendor",
            "test": {}
          }
        }
      }
    }
  }
]",
}
`;

exports[`webpack should create a webpack config with language: language configurations 1`] = `
{
  "js": "[
  {
    "rules": [
      {
        "test": {},
        "type": "asset/resource",
        "generator": {
          "filename": "fonts/[hash][ext][query]"
        }
      },
      {
        "test": {},
        "type": "asset",
        "parser": {
          "dataUrlCondition": {
            "maxSize": 5000
          }
        },
        "generator": {
          "filename": "images/[hash][ext][query]"
        }
      },
      {
        "test": {},
        "type": "asset",
        "parser": {
          "dataUrlCondition": {
            "maxSize": 5000
          }
        },
        "generator": {
          "filename": "images/[hash][ext][query]"
        }
      },
      {
        "test": {},
        "use": [
          "./node_modules/mini-css-extract-plugin/dist/loader.js",
          "css-loader"
        ]
      },
      {
        "test": {},
        "include": [
          "./__fixtures__/src"
        ],
        "resolve": {
          "extensions": [
            ".js",
            ".jsx",
            ".mjs",
            ".cjs"
          ]
        },
        "use": [
          {
            "loader": "babel-loader",
            "options": {
              "presets": [
                "@babel/preset-env"
              ]
            }
          }
        ]
      }
    ]
  }
]",
  "ts": "[
  {
    "rules": [
      {
        "test": {},
        "type": "asset/resource",
        "generator": {
          "filename": "fonts/[hash][ext][query]"
        }
      },
      {
        "test": {},
        "type": "asset",
        "parser": {
          "dataUrlCondition": {
            "maxSize": 5000
          }
        },
        "generator": {
          "filename": "images/[hash][ext][query]"
        }
      },
      {
        "test": {},
        "type": "asset",
        "parser": {
          "dataUrlCondition": {
            "maxSize": 5000
          }
        },
        "generator": {
          "filename": "images/[hash][ext][query]"
        }
      },
      {
        "test": {},
        "use": [
          "./node_modules/mini-css-extract-plugin/dist/loader.js",
          "css-loader"
        ]
      },
      {
        "test": {},
        "include": [
          "./__fixtures__/src"
        ],
        "resolve": {
          "extensions": [
            ".ts",
            ".tsx",
            ".mts",
            ".cts",
            ".js",
            ".jsx",
            ".mjs",
            ".cjs"
          ]
        },
        "use": [
          {
            "loader": "ts-loader"
          }
        ]
      }
    ]
  }
]",
}
`;

exports[`webpack should extend a basic webpack config using external resources: custom configurations 1`] = `
"[
  {
    "host": "localhost",
    "port": 3000,
    "compress": true,
    "historyApiFallback": true,
    "hot": true,
    "devMiddleware": {
      "stats": "errors-only",
      "writeToDisk": true
    },
    "client": {
      "overlay": false,
      "progress": true
    },
    "static": {
      "directory": "./__fixtures__/dist"
    },
    "watchFiles": {
      "paths": [
        "src/**/*"
      ]
    }
  }
]"
`;

exports[`webpack should handle build output bundle errors, stats errors, and stat output: startWpErrorStatsHandler 1`] = `
{
  "errors": [
    [
      "[31mProduction build errors...
basic error[0m",
    ],
    [
      "[31mProduction compile errors...
compile error[0m",
    ],
  ],
  "logs": [
    [
      "[32mStats file created[0m",
    ],
    [
      "[0mStats level [33mundefined[0m...
  lorem: ipsum[0m",
    ],
    [
      "[32mBuild completed[0m",
    ],
  ],
}
`;

exports[`webpack should return specific properties: specific properties 1`] = `
{
  "cleanDist": [Function],
  "createWpConfig": [Function],
  "startWp": [Function],
  "startWpErrorStatsHandler": [Function],
}
`;
